osu.Game/Rulesets/Edit/Checks/CheckAbnormalDifficultySettings.cs

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using osu.Game.Rulesets.Edit.Checks.Components;
+
+namespace osu.Game.Rulesets.Edit.Checks
+{
+    public class CheckAbnormalDifficultySettings : ICheck
+    {
+        public CheckMetadata Metadata => new CheckMetadata(CheckCategory.Settings, "Abnormal difficulty settings");
+
+        public IEnumerable<IssueTemplate> PossibleTemplates => new IssueTemplate[]
+        {
+            new IssueTemplateMoreThanOneDecimal(this),
+            new IssueTemplateOutOfRange(this),
+        };
+
+        public IEnumerable<Issue> Run(BeatmapVerifierContext context)
+        {
+            var diff = context.Beatmap.Difficulty;
+            string ruleset = context.Beatmap.BeatmapInfo.Ruleset.ShortName;
+
+            if (hasMoreThanOneDecimalPlace(diff.ApproachRate))
+                yield return new IssueTemplateMoreThanOneDecimal(this).Create("Approach rate", diff.ApproachRate);
+
+            if (isOutOfRange(diff.ApproachRate))
+                yield return new IssueTemplateOutOfRange(this).Create("Approach rate", diff.ApproachRate);
+
+            if (hasMoreThanOneDecimalPlace(diff.OverallDifficulty))
+                yield return new IssueTemplateMoreThanOneDecimal(this).Create("Overall difficulty", diff.OverallDifficulty);
+
+            if (isOutOfRange(diff.OverallDifficulty))
+                yield return new IssueTemplateOutOfRange(this).Create("Overall difficulty", diff.OverallDifficulty);
+
+            if (hasMoreThanOneDecimalPlace(diff.CircleSize))
+                yield return new IssueTemplateMoreThanOneDecimal(this).Create("Circle size", diff.CircleSize);
+
+            if (ruleset != "mania" && isOutOfRange(diff.CircleSize))
+                yield return new IssueTemplateOutOfRange(this).Create("Circle size", diff.CircleSize);
+
+            if (hasMoreThanOneDecimalPlace(diff.DrainRate))
+                yield return new IssueTemplateMoreThanOneDecimal(this).Create("Drain rate", diff.DrainRate);
+
+            if (isOutOfRange(diff.DrainRate))
+                yield return new IssueTemplateOutOfRange(this).Create("Drain rate", diff.DrainRate);
```

Comment:
I dunno how to feel about these in general. They're presented as "general" settings but they're really not.

- Approach rate does nothing gameplay-relevant in taiko and mania. (It affects conversion RNG in mania, but I wouldn't call that "gameplay-relevant".)
- Drain rate does nothing in taiko.
- Overall difficulty does nothing in catch.

So I don't see why ruleset-specific beatmaps should report all these issues on difficulty attributes they're not even using. The `!= "mania"` check thing is also an eyesore.

If you ask me this should be four checks, per each ruleset. Each ruleset only checks the values it is actively using for gameplay mechanics and ignores the rest. If you wish then make this a base check to reuse the common logic, but the actual details of what ruleset checks what should be up to the ruleset.

Reply:
Also i would like to suggest implementing difficulty specific checks to the settings, for example an Easy difficulty  in standard as the [wiki says](https://osu.ppy.sh/wiki/en/Ranking_criteria/osu!#Difficulty-specific) should match a certain criteria:
  - Approach rate should be 5 or less.
  - Overall difficulty / HP drain rate should be between 1 and 3.
  - Circle size should be 4 or lower.


I think it would be useful to have such a check and in my opinion matches the scope of this PR.

Reply:
Structure seems fine.

> I think it would be useful to have such a check and in my opinion matches the scope of this PR.

I'd prefer to keep that out of this diff, especially if it's not in mapset verifier right now.

Reply:
I think thats it, i will sum up the checks:
- Catch checks for: AR, CS, HP
- Mania checks for: OD, HP
- Osu checks for everything
- Taiko checks for: OD

Please let me know if i missed something.